client/db/bolt/db.go

```diff
@@ -0,0 +1,535 @@
+	}
+	return db.marketOrdersSince(dexB, baseB, quoteB, n, since)
+}
+
+// marketOrdersAll retrieves all orders for the specified DEX and market.
+func (db *boltDB) marketOrdersAll(dexB, baseB, quoteB []byte) ([]*dexdb.MetaOrder, error) {
+	return db.filteredOrders(func(oBkt *bolt.Bucket) bool {
+		return bEqual(dexB, oBkt.Get(dexKey)) && bEqual(baseB, oBkt.Get(baseKey)) &&
+			bEqual(quoteB, oBkt.Get(quoteKey))
+	})
+}
+
+// marketOrdersSince grabs market orders with optional filters for maximum count
+// and age. The sort order is always newest first. If n is 0, there is no limit
+// to the number of orders returned. If using with both n = 0, and since = 0,
+// use marketOrdersAll instead.
+func (db *boltDB) marketOrdersSince(dexB, baseB, quoteB []byte, n int, since uint64) ([]*dexdb.MetaOrder, error) {
+	sinceB := uint64Bytes(since)
+	return db.newestOrders(n, func(oBkt *bolt.Bucket) bool {
+		timeB := oBkt.Get(updateTimeKey)
+		return bEqual(dexB, oBkt.Get(dexKey)) && bEqual(baseB, oBkt.Get(baseKey)) &&
+			bEqual(quoteB, oBkt.Get(quoteKey)) && bytes.Compare(timeB, sinceB) >= 0
+	})
+}
+
+// orderTimePair is used to build an on-the-fly index to sort orders by time.
+type orderTimePair struct {
+	oid []byte
+	t   uint64
+}
+
+// newestOrders returns the n newest orders, filtered with a supplied filter
+// function. Each order's bucket is provided to the filter, and a boolean true
+// return value indicates the order should is eligible to be decoded and
+// returned.
+func (db *boltDB) newestOrders(n int, filter func(*bolt.Bucket) bool) ([]*dexdb.MetaOrder, error) {
+	var orders []*dexdb.MetaOrder
+	db.ordersView(func(master *bolt.Bucket) error {
+		pairs := make([]*orderTimePair, 0, n)
+		master.ForEach(func(oid, _ []byte) error {
+			oBkt := master.Bucket(oid)
+			timeB := oBkt.Get(updateTimeKey)
+			if filter(oBkt) {
+				pairs = append(pairs, &orderTimePair{
+					oid: oid,
+					t:   intCoder.Uint64(timeB),
+				})
+			}
+			return nil
+		})
+		// Sort the pairs newest first.
+		sort.Slice(pairs, func(i, j int) bool { return pairs[i].t > pairs[j].t })
+		if n > 0 && len(pairs) > n {
+			pairs = pairs[:n]
+		}
+
+		for _, pair := range pairs {
+			oBkt := master.Bucket(pair.oid)
+			o, err := decodeOrderBucket(pair.oid, oBkt)
+			if err != nil {
+				return err
+			}
+			orders = append(orders, o)
+		}
+		return nil
+	})
+	return orders, nil
+}
+
+// filteredOrders gets all orders that pass the provided filter function. Each
+// order's bucket is provided to the filter, and a boolean true return value
+// indicates the order should be decoded and returned.
+func (db *boltDB) filteredOrders(filter func(*bolt.Bucket) bool) ([]*dexdb.MetaOrder, error) {
+	var orders []*dexdb.MetaOrder
+	var err error
+	db.ordersView(func(master *bolt.Bucket) error {
+		master.ForEach(func(oid, _ []byte) error {
+			oBkt := master.Bucket(oid)
+			if oBkt == nil {
+				return fmt.Errorf("order %x bucket is not a bucket", oid)
```

Comment:
This error value won't make it to the `err` declared outside this func.  Why not `err := db.ordersView` to be safe?